### PR TITLE
Add support for property notifications (notify signal)

### PR DESCRIPTION
**Pull request by sdroege**
https://github.com/gtk-rs/gir/issues/437

### CODE REVIEW

File: src/analysis/properties.rs [hunk 1]
```diff
@@ -199,7 +217,64 @@ fn analyze_property(
         None
     };
 
-    (getter, setter)
+    let mut used_types: Vec<String> = Vec::with_capacity(4);
+    let trampoline_name = trampolines::analyze(
+        env,
+        &library::Signal {
+            name: format!("notify::{}", name),
+            parameters: Vec::new(),
```
Comment (sdroege) [inline src/analysis/properties.rs]: Somehow we need to tell it to generate a `gpointer` parameter for the `extern "C"` trampoline but ignore it for anything else except for just having it there.

Comment (sdroege) [inline src/analysis/properties.rs]: Done

**Comment by sdroege** [comment]
With those commits I get proper property notification connectors for everything in the gstreamer bindings.

**Comment by sdroege** [comment]
https://github.com/sdroege/gstreamer-rs/commit/74c9f9145a028623187a45ac3c67427c4ad4e508 for how this looks like

**Comment by sdroege** [comment]
@GuillaumeGomez Would be nice to get this into the release, even if *I* don't need it for anything in glib/gtk/etc. But I won't be able to do any fixups for the PR in the next days

**Comment by GuillaumeGomez** [comment]
Hum indeed. Anything against it @EPashkin?

File: src/analysis/ref_mode.rs [hunk 2]
```diff
@@ -27,7 +27,11 @@ impl RefMode {
             ..
         }) = env.config.objects.get(&tid.full_name(&env.library))
         {
-            return ref_mode;
+            if direction == library::ParameterDirection::In {
+                return ref_mode;
+            } else {
+                return RefMode::None;
```
Comment (EPashkin) [inline src/analysis/ref_mode.rs]: Can you give sample, when this needed.

Comment (sdroege) [inline src/analysis/ref_mode.rs]: Caps in gstreamer. Manual but ref-counted, so `ref_mode = "ref"` is configured. And if you look below in that function, the behaviour is the same. It only returns something else than `None` for `In` parameters, as everything else makes no sense.

Without this, e.g. a property returning Caps will return `Option<&Caps>`, which makes no sense and is nothing that could be constructed.

Comment (EPashkin) [inline src/analysis/ref_mode.rs]: if direction== Return { return None } will be cleaner

**Comment by EPashkin** [comment]
I has only one question, but can merge without it solves as this change don't affect gtk-rs.

**Comment by sdroege** [comment]
It should affect every crate that has properties. They will get a notification signal for those properties now.

File: src/analysis/ref_mode.rs (hunk 2 above)
Comment (sdroege) [inline src/analysis/ref_mode.rs]: It would also be for Out parameters (and InOut I guess, but we don't really support them).

**Comment by EPashkin** [comment]
I meant `return None` question. If you can't fix it fast, I merge.

**Comment by sdroege** [comment]
`return None` would return `Option::None` instead of `RefMode::None`. I'm not sure I understand what you mean then :)

File: src/analysis/ref_mode.rs (hunk 2 above)
Comment (EPashkin) [inline src/analysis/ref_mode.rs]:
```
if direction == library::ParameterDirection::Return {
      return RefMode::None;
} else {
      return ref_mode;
}
```

**Comment by EPashkin** [comment]
Ok, then I merge now.
@sdroege Thanks

**Comment by sdroege** [comment]
Thanks :)

File: src/analysis/ref_mode.rs (hunk 2 above)
Comment (sdroege) [inline src/analysis/ref_mode.rs]: It should also be None for ::Out at least

**Comment by sdroege** [comment]
Now only someone needs to regenerate gtk and gio at least, not sure which others have properties

**Comment by EPashkin** [comment]
I will do it tomorrow

**Comment by GuillaumeGomez** [comment]
I'll do gtk this evening, I leave the rest to @EPashkin. ;)

**Comment by EPashkin** [comment]
There path that better added for gtk (I will add it to others)
```diff
--- a/.travis.yml
+++ b/.travis.yml
@@ -15,9 +15,9 @@ script:
   - ./check_init_asserts
   - cargo install clippy || touch clippy_failed
   - if [ ! -f clippy_failed ]; then cargo clippy; fi
-  - cargo test --features embed-lgpl-docs
+  - cargo test --features embed-lgpl-docs --jobs 1
   # catch any sneaked in lgpl docs
-  - cargo build --features purge-lgpl-docs
+  - cargo build --features purge-lgpl-docs --jobs 1
   - git diff -R --exit-code
   - mkdir .cargo
   - echo 'paths = ["."]' > .cargo/config
```

**Comment by GuillaumeGomez** [comment]
Why the `--jobs 1` option btw?

**Comment by EPashkin** [comment]
Your libs take too much memory when compiling.
See https://github.com/gtk-rs/examples/pull/138.

**Comment by GuillaumeGomez** [comment]
Erf. Do you want PRs to add this change or do you prefer I do it?

**Comment by EPashkin** [comment]
As you do regen on gtk, I prefer that you add this fix in your PR

**Comment by GuillaumeGomez** [comment]
Yes sir!

**Pull request merged**